### ISSUE TITLE
Refactor Planning for ADS/IR/V2 work

### ISSUE THREAD

**Issue by rhs** [datawire/ambassador#544]
This is the output of a planning meeting with the goal of figuring out how to proceed on the several refactors we would like to make.

Refactors:

1. IR Refactor
2. Moving to ADS
3. V1 to V2 config
4. Testing

We primarily identified any user requested features we could think of and grouped these into two categories. Features that may significantly inform a given refactor, and features that we don't necessarily expect to inform the refactor, but should function as a litmus test (once the refactor is done, these features should be trivial).

Features:

- Better diagnostics (informs 1, 4):
  + for service developers
  + for ambassador developers [with no access to the ambassador pod]
  + for ops
  + report out of sync pods
- Session affinity (litmus for 1)
- Regex rewriting (litmus for 1)
- SNI
- GRPC UX (litmus for 1)
- Vanity Domain Provisioning
- TCP Proxy (informs 1)
- Faster Reconfig (informs 2)
- Zero Connection Drop (informs 2)
- Multiple Listeners (litmus for 1)
- Endpoint Routing (informs 2)
  + fancy LB algorithms
  + per pod health checking
- Service Discovery Integrations (informs 2)
- Better health stuff (informs 2)
  + outlier rejection
  + ...
- Developer Isolation
- CRDs as UX for existing functionality
- compression

Next Steps:

1. make sure this is complete
2. do a rough prioritization
3. get user stories for all the features that inform any refactors

**Comment by richarddli** [comment]
Here's the list of things we need to tackle once we support v2:

- [ ] Sticky sessions (#422)
- [ ] SNI support / multiple certs (#153)
- [ ] Gzip compression 
- [ ] WebSockets / auth issue (#670)

**Comment by krish512** [comment]
I suggest include Brotli compressions as well

**Comment by Tapppi** [comment]
Any updates on this since 0.50 is GA?

**Comment by richarddli** [comment]
Work is ongoing @Tapppi but PRs are encouraged! We've gotten to 2 on the list.

**Comment by enriched** [comment]
Is gzip support in the works? I see that [envoy has a gzip http filter](https://www.envoyproxy.io/docs/envoy/v1.9.0/api-v2/config/filter/http/gzip/v2/gzip.proto#envoy-api-msg-config-filter-http-gzip-v2-gzip) would there be a way to add this to the configuration manually as a workaround?

**Comment by richarddli** [comment]
@enriched it's straightforward to add now that we have v2 support, and we'd be happy to review the PR (one of our programmers actually wrote that filter). looking for someone in the community to add and test it!

**Comment by kflynn** [comment]
This is long since done (yow). There's a separate issue for gzip support (#744).
